File: Tests/DungeonSoundboardCoreTests/HotkeyTests.swift
import XCTest
@testable import DungeonSoundboardCore

final class HotkeyTests: XCTestCase {
    func testDefaultHotkeysContainExpectedSystemActions() {
        let configuration = HotkeyConfiguration.defaults

        XCTAssertEqual(configuration.hotkey(for: .stopEffects)?.displayText, "Delete")
        XCTAssertNil(configuration.hotkey(for: .stopAll))
        XCTAssertEqual(configuration.hotkey(for: .playPause)?.displayText, "Space")
        XCTAssertEqual(configuration.hotkey(for: .effectsVolumeUp)?.displayText, "Shift++")
        XCTAssertEqual(configuration.hotkey(for: .effectsVolumeDown)?.displayText, "Shift+-")
    }

    func testAssignReportsConflictUnlessResolving() {
        let hotkey = Hotkey(keyCode: 12, label: "Q", modifier: .none)
        var configuration = HotkeyConfiguration(bindings: [
            HotkeyBinding(action: .playPause, hotkey: hotkey)
        ])

        let conflict = configuration.assign(hotkey, to: .stopAll, resolvingConflicts: false)

        XCTAssertEqual(conflict?.existingAction, .playPause)
        XCTAssertEqual(configuration.hotkey(for: .playPause), hotkey)

        let resolved = configuration.assign(hotkey, to: .stopAll, resolvingConflicts: true)

        XCTAssertNil(resolved)
        XCTAssertNil(configuration.hotkey(for: .playPause))
        XCTAssertEqual(configuration.hotkey(for: .stopAll), hotkey)
    }

    func testPhysicalHotkeyIdentityIgnoresKeyboardLayoutLabel() {
        let latin = Hotkey(keyCode: 0, label: "A", modifier: .control)
        let cyrillic = Hotkey(keyCode: 0, label: "Ф", modifier: .control)

        XCTAssertEqual(latin, cyrillic)
        XCTAssertEqual(Set([latin, cyrillic]).count, 1)
    }

    func testConflictDetectionUsesPhysicalKeyInsteadOfLabel() {
        let latin = Hotkey(keyCode: 0, label: "A", modifier: .none)
        let cyrillic = Hotkey(keyCode: 0, label: "Ф", modifier: .none)
        var configuration = HotkeyConfiguration(bindings: [
            HotkeyBinding(action: .playPause, hotkey: latin)
        ])

        let conflict = configuration.assign(cyrillic, to: .stopAll, resolvingConflicts: false)

        XCTAssertEqual(conflict?.existingAction, .playPause)
    }

    func testConfigurationSurvivesCoding() throws {
        let action = HotkeyAction.playMusicTrack(
            playlistID: UUID(uuidString: "00000000-0000-0000-0000-000000000001")!,
            trackID: UUID(uuidString: "00000000-0000-0000-0000-000000000002")!
        )
        let configuration = HotkeyConfiguration(bindings: [
            HotkeyBinding(action: action, hotkey: Hotkey(keyCode: 0, label: "A", modifier: .control))
        ])

        let data = try JSONEncoder().encode(configuration)
        let decoded = try JSONDecoder().decode(HotkeyConfiguration.self, from: data)

        XCTAssertEqual(decoded, configuration)
    }

    func testLegacyDeleteStopAllDefaultMigratesToStopEffects() {
        var configuration = HotkeyConfiguration(bindings: [
            HotkeyBinding(action: .stopAll, hotkey: HotkeyConfiguration.deleteHotkey),
            HotkeyBinding(action: .playPause, hotkey: Hotkey(keyCode: 49, label: "Space", modifier: .none))
        ])

        configuration.migrateLegacyDeleteStopAllDefault()

        XCTAssertNil(configuration.hotkey(for: .stopAll))
        XCTAssertEqual(configuration.hotkey(for: .stopEffects), HotkeyConfiguration.deleteHotkey)
        XCTAssertEqual(configuration.hotkey(for: .playPause)?.displayText, "Space")
    }

    func testHotkeyNormalizationAllowsPlainShiftAndControl() {
        XCTAssertEqual(
            Hotkey.normalized(
                keyCode: 0,
                charactersIgnoringModifiers: "a",
                characters: "a",
                isShiftPressed: false,
                isControlPressed: false,
                isOptionPressed: false,
                isCommandPressed: false
            ),
            Hotkey(keyCode: 0, label: "A", modifier: .none)
        )
        XCTAssertEqual(
            Hotkey.normalized(
                keyCode: 0,
                charactersIgnoringModifiers: "a",
                characters: "A",
                isShiftPressed: true,
                isControlPressed: false,
                isOptionPressed: false,
                isCommandPressed: false
            ),
            Hotkey(keyCode: 0, label: "A", modifier: .shift)
        )
        XCTAssertEqual(
            Hotkey.normalized(
                keyCode: 0,
                charactersIgnoringModifiers: "a",
                characters: "a",
                isShiftPressed: false,
                isControlPressed: true,
                isOptionPressed: false,
                isCommandPressed: false
            ),
            Hotkey(keyCode: 0, label: "A", modifier: .control)
        )
    }

    func testHotkeyNormalizationRejectsUnsupportedModifiers() {
        XCTAssertNil(
            Hotkey.normalized(
                keyCode: 0,
                charactersIgnoringModifiers: "a",
                characters: "a",
                isShiftPressed: false,
                isControlPressed: false,
                isOptionPressed: true,
                isCommandPressed: false
            )
        )
        XCTAssertNil(
            Hotkey.normalized(
                keyCode: 0,
                charactersIgnoringModifiers: "a",
                characters: "a",
                isShiftPressed: false,
                isControlPressed: false,
                isOptionPressed: false,
                isCommandPressed: true
            )
        )
        XCTAssertNil(
            Hotkey.normalized(
                keyCode: 0,
                charactersIgnoringModifiers: "a",
                characters: "a",
                isShiftPressed: true,
                isControlPressed: true,
                isOptionPressed: false,
                isCommandPressed: false
            )
        )
    }

    func testHotkeyNormalizationRejectsTabForNativeFocusTraversal() {
        XCTAssertNil(
            Hotkey.normalized(
                keyCode: 48,
                charactersIgnoringModifiers: "\t",
                characters: "\t",
                isShiftPressed: false,
                isControlPressed: false,
                isOptionPressed: false,
                isCommandPressed: false
            )
        )
    }

    func testHotkeyNormalizationTreatsForwardDeleteAsDelete() {
        XCTAssertEqual(
            Hotkey.normalized(
                keyCode: 117,
                charactersIgnoringModifiers: nil,
                characters: nil,
                isShiftPressed: false,
                isControlPressed: false,
                isOptionPressed: false,
                isCommandPressed: false
            ),
            Hotkey(keyCode: 51, label: "Delete", modifier: .none)
        )
    }

    func testMissingTrackBindingsAreRemoved() {
        let existingPlaylistID = UUID()
        let missingPlaylistID = UUID()
        let existingTrackID = UUID()
        let missingTrackID = UUID()
        var configuration = HotkeyConfiguration(bindings: [
            HotkeyBinding(
                action: .playMusicTrack(playlistID: existingPlaylistID, trackID: existingTrackID),
                hotkey: Hotkey(keyCode: 0, label: "A", modifier: .none)
            ),
            HotkeyBinding(
                action: .playMusicTrack(playlistID: missingPlaylistID, trackID: missingTrackID),
                hotkey: Hotkey(keyCode: 1, label: "S", modifier: .none)
            )
        ])

        configuration.removeMissingTrackBindings(
            musicPlaylistIDs: [existingPlaylistID],
            musicTrackIDs: [existingTrackID],
            effectPlaylistIDs: [],
            effectTrackIDs: []
        )

        XCTAssertEqual(configuration.bindings.count, 1)
        XCTAssertEqual(configuration.bindings.first?.action, .playMusicTrack(playlistID: existingPlaylistID, trackID: existingTrackID))
    }

    func testMoveTransferRetargetsMusicBindingAndKeepsHotkey() {
        let sourcePlaylistID = UUID()
        let destinationPlaylistID = UUID()
        let trackID = UUID()
        let hotkey = Hotkey(keyCode: 12, label: "Q", modifier: .control)
        var configuration = HotkeyConfiguration(bindings: [
            HotkeyBinding(
                action: .playMusicTrack(playlistID: sourcePlaylistID, trackID: trackID),
                hotkey: hotkey
            )
        ])
        let result = TrackTransferResult(
            operation: .move,
            role: .music,
            sourcePlaylistID: sourcePlaylistID,
            destinationPlaylistID: destinationPlaylistID,
            transferred: [TrackTransferRecord(sourceTrackID: trackID, destinationTrackID: trackID)],
            duplicateTrackIDs: [],
            missingTrackIDs: []
        )

        XCTAssertEqual(configuration.applyTransferResult(result), 1)
        XCTAssertNil(
            configuration.hotkey(
                for: .playMusicTrack(playlistID: sourcePlaylistID, trackID: trackID)
            )
        )
        XCTAssertEqual(
            configuration.hotkey(
                for: .playMusicTrack(playlistID: destinationPlaylistID, trackID: trackID)
            ),
            hotkey
        )
    }

    func testCopyTransferDoesNotRetargetOriginalBinding() {
        let sourcePlaylistID = UUID()
        let destinationPlaylistID = UUID()
        let sourceTrackID = UUID()
        let copiedTrackID = UUID()
        let action = HotkeyAction.playEffect(playlistID: sourcePlaylistID, trackID: sourceTrackID)
        let hotkey = Hotkey(keyCode: 13, label: "W", modifier: .none)
        var configuration = HotkeyConfiguration(bindings: [
            HotkeyBinding(action: action, hotkey: hotkey)
        ])
        let result = TrackTransferResult(
            operation: .copy,
            role: .effect,
            sourcePlaylistID: sourcePlaylistID,
            destinationPlaylistID: destinationPlaylistID,
            transferred: [
                TrackTransferRecord(sourceTrackID: sourceTrackID, destinationTrackID: copiedTrackID)
            ],
            duplicateTrackIDs: [],
            missingTrackIDs: []
        )

        XCTAssertEqual(configuration.applyTransferResult(result), 0)
        XCTAssertEqual(configuration.hotkey(for: action), hotkey)
    }

    func testReconcileRetargetsUniquelyMovedTrackAndRemovesMissingTrack() {
        let oldPlaylistID = UUID()
        let destinationPlaylistID = UUID()
        let movedTrackID = UUID()
        let missingTrackID = UUID()
        let movedHotkey = Hotkey(keyCode: 14, label: "E", modifier: .none)
        var configuration = HotkeyConfiguration(bindings: [
            HotkeyBinding(
                action: .playMusicTrack(playlistID: oldPlaylistID, trackID: movedTrackID),
                hotkey: movedHotkey
            ),
            HotkeyBinding(
                action: .playMusicTrack(playlistID: oldPlaylistID, trackID: missingTrackID),
                hotkey: Hotkey(keyCode: 15, label: "R", modifier: .none)
            ),
            HotkeyBinding(
                action: .playPause,
                hotkey: Hotkey(keyCode: 49, label: "Space", modifier: .none)
            )
        ])
        let musicPlaylists = [
            Playlist(
                id: destinationPlaylistID,
                name: "Destination",
                tracks: [Track(id: movedTrackID, title: "Moved", path: "/tmp/moved.mp3")]
            )
        ]

        let result = configuration.reconcileTrackBindings(
            musicPlaylists: musicPlaylists,
            effectPlaylists: []
        )

        XCTAssertEqual(result, HotkeyReconciliationResult(retargetedCount: 1, removedCount: 1))
        XCTAssertEqual(
            configuration.hotkey(
                for: .playMusicTrack(playlistID: destinationPlaylistID, trackID: movedTrackID)
            ),
            movedHotkey
        )
        XCTAssertNotNil(configuration.hotkey(for: .playPause))
    }

    func testReconcileKeepsValidExactPairWhenTrackIDHasMultipleOwners() {
        let firstPlaylistID = UUID()
        let secondPlaylistID = UUID()
        let trackID = UUID()
        let hotkey = Hotkey(keyCode: 16, label: "Y", modifier: .none)
        var configuration = HotkeyConfiguration(bindings: [
            HotkeyBinding(
                action: .playMusicTrack(playlistID: firstPlaylistID, trackID: trackID),
                hotkey: hotkey
            )
        ])
        let duplicatedTrack = Track(id: trackID, title: "Duplicate ID", path: "/tmp/a.mp3")

        let result = configuration.reconcileTrackBindings(
            musicPlaylists: [
                Playlist(id: firstPlaylistID, name: "First", tracks: [duplicatedTrack]),
                Playlist(id: secondPlaylistID, name: "Second", tracks: [duplicatedTrack])
            ],
            effectPlaylists: []
        )

        XCTAssertFalse(result.didChange)
        XCTAssertEqual(
            configuration.hotkey(
                for: .playMusicTrack(playlistID: firstPlaylistID, trackID: trackID)
            ),
            hotkey
        )
    }

    func testReconcileRemovesAmbiguousBindingWhenOriginalPairIsGone() {
        let oldPlaylistID = UUID()
        let firstOwnerID = UUID()
        let secondOwnerID = UUID()
        let trackID = UUID()
        var configuration = HotkeyConfiguration(bindings: [
            HotkeyBinding(
                action: .playEffect(playlistID: oldPlaylistID, trackID: trackID),
                hotkey: Hotkey(keyCode: 17, label: "T", modifier: .none)
            )
        ])
        let duplicatedTrack = Track(id: trackID, title: "Duplicate ID", path: "/tmp/effect.wav", role: .effect)

        let result = configuration.reconcileTrackBindings(
            musicPlaylists: [],
            effectPlaylists: [
                EffectPlaylist(id: firstOwnerID, name: "First", effects: [duplicatedTrack]),
                EffectPlaylist(id: secondOwnerID, name: "Second", effects: [duplicatedTrack])
            ]
        )

        XCTAssertEqual(result.removedCount, 1)
        XCTAssertTrue(configuration.bindings.isEmpty)
    }
}

File: windows/tests/DungeonSoundboard.Core.Tests/HotkeyTests.cs
using System.Text.Json;
using DungeonSoundboard.Core.Models;
using DungeonSoundboard.Core.Serialization;
using Xunit;

namespace DungeonSoundboard.Core.Tests;

public sealed class HotkeyTests
{
    [Fact]
    public void HotkeyNormalizationRejectsTab()
    {
        Assert.Null(Hotkey.Normalized(
            HotkeyConfiguration.TabKeyCode,
            "\t",
            "\t",
            isShiftPressed: false,
            isControlPressed: false,
            isOptionPressed: false,
            isCommandPressed: false));
    }

    [Fact]
    public void DefaultHotkeysContainExpectedSystemActions()
    {
        var configuration = HotkeyConfiguration.Defaults;

        Assert.Equal("Delete", configuration.HotkeyFor(HotkeyAction.StopEffects)?.DisplayText);
        Assert.Null(configuration.HotkeyFor(HotkeyAction.StopAll));
        Assert.Equal("Space", configuration.HotkeyFor(HotkeyAction.PlayPause)?.DisplayText);
        Assert.Equal("Shift++", configuration.HotkeyFor(HotkeyAction.EffectsVolumeUp)?.DisplayText);
        Assert.Equal("Shift+-", configuration.HotkeyFor(HotkeyAction.EffectsVolumeDown)?.DisplayText);
    }

    [Fact]
    public void RestoreDefaultRestoresOnlyTheRequestedSystemBinding()
    {
        var customPlayPause = new Hotkey(0, "A", HotkeyModifier.None);
        var customStopEffects = new Hotkey(1, "B", HotkeyModifier.None);
        var configuration = new HotkeyConfiguration(
        [
            new HotkeyBinding(HotkeyAction.PlayPause, customPlayPause),
            new HotkeyBinding(HotkeyAction.StopEffects, customStopEffects)
        ]);

        configuration.RestoreDefault(HotkeyAction.PlayPause);

        Assert.Equal("Space", configuration.HotkeyFor(HotkeyAction.PlayPause)?.DisplayText);
        Assert.Equal(customStopEffects, configuration.HotkeyFor(HotkeyAction.StopEffects));
    }

    [Fact]
    public void RestoreDefaultSystemBindingRemovesAConflictingTrackBinding()
    {
        var playlistId = Guid.NewGuid();
        var trackId = Guid.NewGuid();
        var configuration = new HotkeyConfiguration(
        [
            new HotkeyBinding(HotkeyAction.PlayMusicTrack(playlistId, trackId), new Hotkey(HotkeyConfiguration.SpaceKeyCode, "Space", HotkeyModifier.None))
        ]);

        configuration.RestoreDefault(HotkeyAction.PlayPause);

        Assert.Equal("Space", configuration.HotkeyFor(HotkeyAction.PlayPause)?.DisplayText);
        Assert.Null(configuration.HotkeyFor(HotkeyAction.PlayMusicTrack(playlistId, trackId)));
    }

    [Fact]
    public void RestoreDefaultClearsAnActionThatIsUnassignedByDefault()
    {
        var configuration = new HotkeyConfiguration(
        [
            new HotkeyBinding(HotkeyAction.StopAll, new Hotkey(0, "A", HotkeyModifier.None))
        ]);

        configuration.RestoreDefault(HotkeyAction.StopAll);

        Assert.Null(configuration.HotkeyFor(HotkeyAction.StopAll));
    }

    [Fact]
    public void AssignReportsConflictUnlessResolving()
    {
        var hotkey = new Hotkey(12, "Q", HotkeyModifier.None);
        var configuration = new HotkeyConfiguration(
        [
            new HotkeyBinding(HotkeyAction.PlayPause, hotkey)
        ]);

        var conflict = configuration.Assign(hotkey, HotkeyAction.StopAll, resolvingConflicts: false);

        Assert.Equal(HotkeyAction.PlayPause, conflict?.ExistingAction);
        Assert.Equal(hotkey, configuration.HotkeyFor(HotkeyAction.PlayPause));

        var resolved = configuration.Assign(hotkey, HotkeyAction.StopAll, resolvingConflicts: true);

        Assert.Null(resolved);
        Assert.Null(configuration.HotkeyFor(HotkeyAction.PlayPause));
        Assert.Equal(hotkey, configuration.HotkeyFor(HotkeyAction.StopAll));
    }

    [Fact]
    public void ConfigurationSurvivesCoding()
    {
        var action = HotkeyAction.PlayMusicTrack(
            Guid.Parse("00000000-0000-0000-0000-000000000001"),
            Guid.Parse("00000000-0000-0000-0000-000000000002"));
        var configuration = new HotkeyConfiguration(
        [
            new HotkeyBinding(action, new Hotkey(0, "A", HotkeyModifier.Control))
        ]);

        var data = JsonSerializer.SerializeToUtf8Bytes(configuration, JsonDefaults.Options);
        var decoded = JsonSerializer.Deserialize<HotkeyConfiguration>(data, JsonDefaults.Options);

        Assert.Equal(configuration, decoded);
    }

    [Fact]
    public void LegacyDeleteStopAllDefaultMigratesToStopEffects()
    {
        var configuration = new HotkeyConfiguration(
        [
            new HotkeyBinding(HotkeyAction.StopAll, HotkeyConfiguration.DeleteHotkey),
            new HotkeyBinding(HotkeyAction.PlayPause, new Hotkey(49, "Space", HotkeyModifier.None))
        ]);

        configuration.MigrateLegacyDeleteStopAllDefault();

        Assert.Null(configuration.HotkeyFor(HotkeyAction.StopAll));
        Assert.Equal(HotkeyConfiguration.DeleteHotkey, configuration.HotkeyFor(HotkeyAction.StopEffects));
        Assert.Equal("Space", configuration.HotkeyFor(HotkeyAction.PlayPause)?.DisplayText);
    }

    [Fact]
    public void HotkeyNormalizationAllowsPlainShiftAndControl()
    {
        Assert.Equal(
            new Hotkey(0, "A", HotkeyModifier.None),
            Hotkey.Normalized(0, "a", "a", false, false, false, false));
        Assert.Equal(
            new Hotkey(0, "A", HotkeyModifier.Shift),
            Hotkey.Normalized(0, "a", "A", true, false, false, false));
        Assert.Equal(
            new Hotkey(0, "A", HotkeyModifier.Control),
            Hotkey.Normalized(0, "a", "a", false, true, false, false));
    }

    [Fact]
    public void HotkeyNormalizationRejectsUnsupportedModifiers()
    {
        Assert.Null(Hotkey.Normalized(0, "a", "a", false, false, true, false));
        Assert.Null(Hotkey.Normalized(0, "a", "a", false, false, false, true));
        Assert.Null(Hotkey.Normalized(0, "a", "a", true, true, false, false));
    }

    [Fact]
    public void HotkeyNormalizationTreatsForwardDeleteAsDelete()
    {
        Assert.Equal(
            new Hotkey(51, "Delete", HotkeyModifier.None),
            Hotkey.Normalized(117, null, null, false, false, false, false));
    }

    [Fact]
    public void MissingTrackBindingsAreRemoved()
    {
        var existingPlaylistId = Guid.NewGuid();
        var missingPlaylistId = Guid.NewGuid();
        var existingTrackId = Guid.NewGuid();
        var missingTrackId = Guid.NewGuid();
        var configuration = new HotkeyConfiguration(
        [
            new HotkeyBinding(
                HotkeyAction.PlayMusicTrack(existingPlaylistId, existingTrackId),
                new Hotkey(0, "A", HotkeyModifier.None)),
            new HotkeyBinding(
                HotkeyAction.PlayMusicTrack(missingPlaylistId, missingTrackId),
                new Hotkey(1, "S", HotkeyModifier.None))
        ]);

        configuration.RemoveMissingTrackBindings(
            new HashSet<Guid> { existingPlaylistId },
            new HashSet<Guid> { existingTrackId },
            new HashSet<Guid>(),
            new HashSet<Guid>());

        Assert.Single(configuration.Bindings);
        Assert.Equal(HotkeyAction.PlayMusicTrack(existingPlaylistId, existingTrackId), configuration.Bindings[0].Action);
    }
}
